### PR TITLE
Deactivated users can be reactivated

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change
Deactivated users can be reactivated
Add deactivated users tab to manage users

## Link to relevant ticket
[CRIMRE-349](https://dsdmoj.atlassian.net/browse/CRIMRE-349)

### Before changes:
Screenshot 2023-06-07 at 16 33 10

### After changes:
Screenshot 2023-06-07 at 16 31 46
Screenshot 2023-06-07 at 16 31 37
Screenshot 2023-06-07 at 16 31 28

## How to manually test the feature

[CRIMRE-349]: https://dsdmoj.atlassian.net/browse/CRIMRE-349?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/controllers/admin/manage_users/deactivated_users_controller.rb [hunk 1]
```diff
@@ -1,22 +1,32 @@
...
+        user = User.active.find(params[:id])
         user.deactivate!
 
         set_flash :user_deactivated, user_name: user.name
         redirect_to admin_manage_users_root_path
       end
 
-      private
+      def confirm_reactivate
+        @user = User.find(params[:id])
```
Comment (willmcb) [inline app/controllers/admin/manage_users/deactivated_users_controller.rb:21]: Should this not be within the 'active' users scpe as per line 13?

File: spec/system/manage_users/reactivate_a_user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  include_context 'with an existing user'
+  let(:confirm_path) { confirm_reactivate_admin_manage_users_deactivated_user_path(deactivated_user) }
+  let(:deactivated_user) { active_user }
+
+  before do
+    active_user.deactivate!
+    visit admin_manage_users_deactivated_users_path
+  end
+
+  describe 'reactivating a deactivated users' do
```
Comment (willmcb) [inline spec/system/manage_users/reactivate_a_user_spec.rb:14]: users --> user typo 👍

File: spec/system/manage_users/reactivate_a_user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    it 'prompts to confirm the action' do
+      expect(page).to have_content(
+        "Are you sure you want to reactivate Zoe Blogs's account?"
+      )
+    end
+
+    describe 'clicking on "Yes, reactivate"' do
+      it 'redirects to the manage user list' do
```
Comment (willmcb) [inline spec/system/manage_users/reactivate_a_user_spec.rb:28]: manage user --> manage users

**Review by willmcb — CHANGES_REQUESTED** [review]
Only blocking thing is the comment about using the `active` scope to check reactivation. I may be wrong about it but it didn't make sense to as it atm.

Otherwise looks good - left some other comments to think about - which I am also happy to chat through with you tomorrow 😃

File: app/models/user.rb [hunk 4]
```diff
@@ -39,6 +43,10 @@ def deactivate!
     update!(deactivated_at: Time.zone.now)
   end
 
+  def reactivate!
+    update!(deactivated_at: nil)
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/models/user.rb:47]: Yes, I agree. I think we should log all these user management events.

File: config/locales/en/manage_users.yml [hunk 5]
```diff
@@ -78,3 +86,5 @@ en:
       deactivated_users:
         new:
           page_title: Confirm deactivation
+        confirm_reactivate:
+          page_title: Confirm reactivate
```
[the remark this answers is not in the document]
Comment (arthurashman) [reply config/locales/en/manage_users.yml:90]: +1

File: app/controllers/admin/manage_users/deactivated_users_controller.rb (hunk 1 above)
Comment (timpeat) [reply app/controllers/admin/manage_users/deactivated_users_controller.rb:21]: good spot.

**Review by willmcb — APPROVED** [review]
👍

File: app/models/user.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (timpeat) [reply app/models/user.rb:57]: Ticket for use authorisation history: [CRIMRE-360](https://dsdmoj.atlassian.net/browse/CRIMRE-360)

File: app/models/user.rb [hunk 6]
```diff
@@ -21,6 +21,10 @@ class CannotRenewIfActive < StandardError; end
     where('auth_subject_id IS NOT NULL AND deactivated_at IS NULL')
   }
 
+  scope :deactivated, lambda {
+    where('auth_subject_id IS NOT NULL AND deactivated_at IS NOT NULL')
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/models/user.rb:27]: Deactivated is means previously "activated" then "deactivated". Inverse of "activated" is not yet activated. CRIMRE-360 will help clarify this I hope.

File: app/views/admin/manage_users/deactivated_users/confirm_reactivate.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<% title t('.page_title') %>
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-full">
+    <h1 class="govuk-heading-xl govuk-!-margin-bottom-3">
+      <%= confirm_text(:reactivate_user, user_name: @user.name) %>
+    </h1>
+  </div>
+</div>
+
+<%= form_with(url: reactivate_admin_manage_users_deactivated_user_path(@user), method: :patch) do |f| %>
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/views/admin/manage_users/deactivated_users/confirm_reactivate.html.erb:10]: Possibly. I'd quite like to loose the "admin" namespace altogether.

**Pull request merged**